src/async_fuse/memfs/s3_metadata.rs

```diff
@@ -293,57 +279,50 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
         //         flags,
         //     ))
         // } else {
-        let result = node.dup_fd(o_flags).await;
-        self.set_node_to_kv_engine(ino, node).await?;
-        result
+        retry_txn!(TXN_RETRY_LIMIT, {
+            let mut txn = self.kv_engine.new_meta_txn().await;
+            let node = self.get_inode_from_txn(txn.as_mut(), ino).await?;
+            let o_flags = fs_util::parse_oflag(flags);
+            node.open_pre_check(o_flags, context.user_id, context.group_id)?;
+
+            let result = node.dup_fd(o_flags).await;
+            txn.set(
+                &KeyType::INum2Node(ino),
+                &ValueType::Node(node.into_serial_node()),
+            );
+            (txn.commit().await, result)
+        })?
     }
 
     #[instrument(skip(self), err, ret)]
     async fn getattr(&self, ino: u64) -> DatenLordResult<(Duration, FuseAttr)> {
-        let inode_wrap = self
+        let inode = self
             .get_node_from_kv_engine(ino)
             .await?
             .ok_or_else(|| build_inconsistent_fs!(ino))?;
-
-        let inode = inode_wrap;
         let attr = inode.get_attr();
-        debug!(
-            "getattr() cache hit when searching the attribute of ino={} and name={:?}",
-            ino,
-            inode.get_name(),
-        );
         let ttl = Duration::new(MY_TTL_SEC, 0);
         let fuse_attr = fs_util::convert_to_fuse_attr(attr);
         Ok((ttl, fuse_attr))
     }
 
     #[instrument(skip(self))]
     async fn forget(&self, ino: u64, nlookup: u64) -> DatenLordResult<()> {
-        let current_count: i64;
-        {
-            let inode = self
-                .get_node_from_kv_engine(ino)
-                .await?
-                .ok_or_else(|| build_inconsistent_fs!(ino))?;
-
-            let previous_count = inode.dec_lookup_count_by(nlookup);
-            current_count = inode.get_lookup_count();
-            debug_assert!(current_count >= 0);
-            debug_assert_eq!(
-                previous_count.overflow_sub(current_count),
-                nlookup.cast::<i64>()
-            ); // assert no race forget
-            debug!(
-                "forget() successfully reduced lookup count of ino={} and name={:?} from {} to {}",
-                ino,
-                inode.get_name(),
-                previous_count,
-                current_count,
-            );
-            self.set_node_to_kv_engine(ino, inode).await?;
-        };
-        self.try_delete_node(ino).await?;
-
+        retry_txn!(TXN_RETRY_LIMIT, {
+            let mut txn = self.kv_engine.new_meta_txn().await;
```

Comment:
This function call is always there, shall we put this statement into the retry_txn macro?

Reply:
I will keep this statement. It can enhance readability. Meanwhile, the semantics of `retry_txn` is just to retry a transaction, and I think this way is better.